checks/connection_filesystem.go

```diff
@@ -0,0 +1,367 @@
+package checks
+
+import (
+	"bytes"
+	gocontext "context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	gcs "cloud.google.com/go/storage"
+	s3Types "github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	awsUtil "github.com/flanksource/canary-checker/pkg/clients/aws"
+	gcpUtil "github.com/flanksource/canary-checker/pkg/clients/gcp"
+	"github.com/flanksource/canary-checker/pkg/utils"
+	"github.com/flanksource/duty/connection"
+	"github.com/flanksource/duty/context"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/duty/types"
+	"github.com/hirochachacha/go-smb2"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// s3FS implements connection.Filesystem for S3
+type s3FS struct {
+	*s3.Client
+	Bucket string
+}
+
+func newS3FS(ctx *context.Context, bucket string, conn connection.AWSConnection) (*s3FS, error) {
+	cfg, err := awsUtil.NewSession(ctx, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &s3FS{
+		Client: s3.NewFromConfig(*cfg, func(o *s3.Options) {
+			o.UsePathStyle = conn.UsePathStyle
+		}),
+		Bucket: strings.TrimPrefix(bucket, "s3://"),
+	}
+
+	return client, nil
+}
+
+func (t *s3FS) Close() error {
+	return nil // NOOP
+}
+
+func (t *s3FS) ReadDir(name string) ([]os.FileInfo, error) {
+	req := &s3.ListObjectsInput{
+		Bucket: aws.String(name),
+	}
+	resp, err := t.Client.ListObjects(gocontext.TODO(), req)
+	if err != nil {
+		return nil, err
+	}
+
+	var output []os.FileInfo
+	for _, r := range resp.Contents {
+		output = append(output, &awsUtil.S3FileInfo{Object: r})
+	}
+
+	return output, nil
+}
+
+func (t *s3FS) Stat(path string) (os.FileInfo, error) {
+	headObject, err := t.Client.HeadObject(gocontext.TODO(), &s3.HeadObjectInput{
+		Bucket: aws.String(t.Bucket),
+		Key:    aws.String(path),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	fileInfo := &awsUtil.S3FileInfo{
+		Object: s3Types.Object{
+			Key:          utils.Ptr(filepath.Base(path)),
+			Size:         headObject.ContentLength,
+			LastModified: headObject.LastModified,
+			ETag:         headObject.ETag,
+		},
+	}
+
+	return fileInfo, nil
+}
+
+func (t *s3FS) Read(ctx gocontext.Context, key string) (io.ReadCloser, error) {
+	results, err := t.Client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(t.Bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer results.Body.Close()
+
+	return results.Body, nil
+}
+
+func (t *s3FS) Write(ctx gocontext.Context, path string, data []byte) (os.FileInfo, error) {
```

Comment:
I don't think we should use []byte as it will consume a lot of memory with large files - rather use io.Reader as the interface and an io.TeeReader with a sha checksum (which is also an io.Writer)  to calc the checksum